[P013_HCSR04] added error checking and logging

JojoS62 opened the pull request:
there were several errors that all resulted in a distance value reading of 0. Added an errorState that is printed to the log and helps troubleshooting.
Max. Distance checking can be turned off by setting max distance to 0 cm. If max. distance is used, exceeding the range will return a distance of 0 cm and print this to the log.

JojoS62 committed:
added error checking and logging

there were several errors that all resulted in a distance value reading of 0. Added an errorState that is printed to the log and helps troubleshooting.
Max. Distance checking can be turned off by setting max distance to 0 cm. If max. distance is used, exceeding the range will return a distance of 0 cm and print this to the log.

TD-er commented:
Is there also a minimum distance this sensor can measure?

ShardanX commented:
Hello Gijs,
the datasheet of the HC-SR04 says 2cm - 400cm. From my experience 400 cm might be optimistic or my cheap sensors are lower quality :)  More realistic are 300..350 cm max.
[HC-SR04Users_Manual.pdf](https://github.com/letscontrolit/ESPEasy/files/2431821/HC-SR04Users_Manual.pdf)

TD-er commented:
Speed of sound is about 3 msec per meter, so a ping over 1 meter takes 6 msec.
1 cm = 60 usec. => 2cm = 120 usec.

Is ESP8266 able to react in such a short time?
Is the pulse short enough?
I wouldn't be surprised if reliable detection within 10 - 20 cm is near impossible.

ShardanX commented:
I'll test with newest release

ShardanX commented:
For my test circut the minimum is around 3 cm.... not too reliable, sometimes the reading glitches to 4cm.. lower distances lead to wrong readings of 8...9 cm.
![hc-sr04_circuit](https://user-images.githubusercontent.com/28200539/46261222-c4ef1c80-c4f0-11e8-94e7-ed39e491b3d8.jpg)
![hc-sr04_test](https://user-images.githubusercontent.com/28200539/46261224-c882a380-c4f0-11e8-997f-fcd4f77d5d25.jpg)

TD-er commented:
Maybe this minimum should also be added as a note to the plugin?

ShardanX commented:
Good idea i think

TD-er commented:
One question about this last commit.
You're making changes in the external lib. How will that affect builds made with Arduino IDE?
If this code is copied, there is no issue I guess, but if it is installed via the library manager of Arduino, is it then causing issues?
Maybe these changes should also be offered to the original library?

JojoS62 commented:
I took care that the functions are compatible the original version. And I'm using platformio, installed after the wiki description (which is really good!).
So to be honest, I don't know the side effects with the Arduino boardmanager. But the code for the lib is included in ESPEasy and the lib is renamed from the original. There are already changes that are not in the original lib. The original is also for ATTinys where every more byte in code size can cause a problem, so I don't know if the modifications are welcome.

JojoS62 commented:
This comment was in #1575 but belongs here:
- I get errors messages now in the log when the trigger or echo pin is not connected (Wiki for this sensor should be extended)
- max. distance = 0: no distance checking, returns max. distance that the sensor can send
- max. distance >0: error message in log, value output is set to 0.0 cm
- in state mode: errors are logged when sensor read fails (but logging freq is 10 per second)

? Log messages like cyclic distance reading are at INFO level. Should this be changed to DEBUG or is this the standard for other modules also?

TD-er commented:
As a side note, before merge this has to be checked also.
@LisaM did apparently also suggest a change, a long time ago.
See [this post on the forum](https://www.letscontrolit.com/forum/viewtopic.php?f=5&t=3432&hilit=HC+SR04+update&start=10#p18413)

Maybe she also did add something that could be useful here?

JojoS62 commented:
about a minimum distance:
It wouldn't be difficult to add a minimum setting, but the sensor cannot safely check a minimum. If the sensor is covered or dirty, it may response a long distance.
I think it is also better to do value checking on the controller level because there the limit are easier to change.

JojoS62 commented:
> Maybe she also did add something that could be useful here?
I will check it, as far as I understand it is about using the median filtering. That is useful and already implemented in the NewPing lib.

JojoS62 committed:
added float calculation for time to distance conversion

returns higher resolution, especially for measuring in inches the integer resolution was low.

JojoS62 committed:
added more settings: unit, filtertype and size

Measuring unit can be changed bewtween metric/imperial
Filtertype can none or median. If median a filtersize can be set.
Meauring value time to distance conversion now as float division, resulting in a higher resolution output.

JojoS62 commented:
*** do not merge, some more fixes are necessary ***

The new settings are working. I have only a problem with setting the defaults. I tried to set defaults in PLUGIN_DEVICE_ADD, but this has no effect.
A valid setting for filtersize > 0 is necessary if median filter is set, otherwise no measuring value is returned.

![grafik](https://user-images.githubusercontent.com/6775562/46357330-ed982300-c664-11e8-8acd-7718be7e037c.png)

sample log:
1202007 : ULTRASONIC : TaskNr: 10 Distance: 87.54
1203007 : ULTRASONIC : TaskNr: 10 Distance: 87.56
1204007 : ULTRASONIC : TaskNr: 10 Distance: 87.91
1205007 : ULTRASONIC : TaskNr: 10 Distance: 87.93
1206007 : ULTRASONIC : TaskNr: 10 Distance: 87.95
1207007 : ULTRASONIC : TaskNr: 10 Distance: 87.51
1208007 : ULTRASONIC : TaskNr: 10 Distance: 87.54

JojoS62 committed:
max distance depends on measurement unit

setting is in user units, ping function needs cm. Conversion added.
added unit to log output.

JojoS62 committed:
code clean up

added default for filter size = 5.
clode clieanup: using constants, read setting into local vars for better readability,

JojoS62 committed:
fixed 'unused variable' warning

JojoS62 committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into mega

uzi18 commented:
For me all these sensors works bad with too neer or to far measurment.
Maybe better to set NaN if distance is unreliable, or out of range?

JojoS62 commented:
I don't know if NaN is a supported value, is it handled in other Plugins also like this? Also the controllers must handle the NaN correctly.
But the US sensors have some limits by design, if there is an detectable error the distance value is set to 0, so this an indicator for invalid measurement.

JojoS62 commented:
I'm happy with the result now. I have tested also various combinations of settings and they work as expected, there is also no memory leak.
The switching between metric/imperial was not necessary, it is possible also with the standard conversion capability in ESPEasy. This can be used also to convert the cm result to a mm or m reading.
From my side the update is ready to merge.
